### PR TITLE
feat(Table): implement table component in Design System

### PR DESCRIPTION
**Pull request by arturfracala**
Resolves: #{issue-number}

## Description

The `Table` component is a customizable solution for displaying tabular data. It supports features like sorting, column resizing, row selection, and striped row styling.

## Storybook

https://feature-DPS-4377--613a8e945a5665003a05113b.chromatic.com

## Checklist

**Obligatory:**

- [ ] Self review (use this as your final check for proposed changes before requesting the review)
- [ ] Add correct label
- [ ] Assign pull request with the correct issue

## Release Notes

- **New Features**
  - Introduced a versatile `Table` component with support for sorting, selection, and resizable columns.
  - Added `TableHeader`, `TableBody`, and `TableRow` components for structured table rendering.
  - Implemented responsive styling for different table sizes and interactive elements.
  - Added a custom hook `useTable` to manage row selection logic.
  - Introduced a new `MultiSelect` story for selecting multiple rows in Storybook.

- **Styling Enhancements**
  - Added comprehensive styles for table components, including hover effects and alternating row colors.
  - Introduced mixins for consistent sizing, pinning styles, and striping effects.

- **Documentation**
  - Created a new documentation file for the `Table` component, detailing its features and usage examples.
  - Added Storybook stories to showcase various configurations and functionalities of the `Table` component.

- **Type Definitions**
  - Added TypeScript interfaces and types for better type safety and component props management.

### CODE REVIEW

**Comment by vladko-uxds** [comment]
Great work!
Few things for fix:

- [ ] Header height for all size -> `36px`
- [ ] Header font size -` paragraph -sm-bold`
- [ ] Header: spacing between label and sorting icon -> `8px` (token)
- [ ] Cell sizes, heights: `Small - 48px`, `Medium-56px`, `Large-64px `
- [ ] Cell text style: -paragraph-md
- [ ] Pinning: shadow as token -> Pinned header `--shadow-fixed-bottom`(token), Left column (`--shadow-fixed-right'), Right column (`--shadow-fixed-left`)

**Comment by vladko-uxds** [comment]
Hover on sorting icon in some columns make shifts

Uploading Monosnap screencast 2024-11-27 10-39-53.mp4…

**Comment by vladko-uxds** [comment]
Resizing affects both sides instead of only the side being adjusted, and it also triggers changes in the widths of other columns.

https://github.com/user-attachments/assets/fac5d312-3e40-48a5-a993-1e4139e4a0af

**Review by VadymBezpalko — APPROVED** [review]

**Review by VadymBezpalko — COMMENTED** [review]
Great job! 👏  I approve this Pull Request, but please pay attention to Vlad's comments regarding the UX part. And please check the comments from CodeRabbit as well 🐰.

**Comment by vladko-uxds** [comment]
Add please properties to turn on "Stripped” Rows or Columns like on example:

image
image

So the user of component can turn "Stripped " props for a specific rows or columns

**Comment by arturfracala** [comment]
> Add please properties to turn on "Stripped” Rows or Columns like on example:
> 
> image image
> 
> So the user of component can turn "Stripped " props for a specific rows or columns

@vladko-uxds it's done. Could you check it?

File: packages/react-components/src/components/Table/Table.mdx [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+import * as Stories from './Table.stories';
+
+<Title>Table</Title>
+
+[Intro](#Intro) | [Component API](#ComponentAPI) | [Content Spec](#ContentSpec)
+
+## Intro <a id="Intro" />
+
+`Table` component is a flexible and interactive table designed for displaying structured data. It supports key features such as sorting, column resizing, row selection, and a customizable layout.
```
[the remark this answers is not in the document]
Comment (marcinsawicki) [reply packages/react-components/src/components/Table/Table.mdx:11]: Unnecessary

File: packages/react-components/src/components/Table/Table.mdx [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+`Table` component is a flexible and interactive table designed for displaying structured data. It supports key features such as sorting, column resizing, row selection, and a customizable layout. 
+
+
+<Canvas of={Stories.Default} sourceState="none" />
+
+#### Example implementation
+
+```jsx
+
```
Comment (marcinsawicki) [inline packages/react-components/src/components/Table/Table.mdx:19]: Please add imports for components used in this example

File: packages/react-components/src/components/Table/Table.mdx [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+ const getRowId = (row: Data) => row.id;
+
+  return (
+    <Table size="medium" data={data} columns={columns} getRowId={getRowId} />
+  );
+```
+
+## Component API <a id="ComponentAPI" />
+
+<ArgTypes of={Stories.Default}  />
```
[the remark this answers is not in the document]
Comment (marcinsawicki) [reply packages/react-components/src/components/Table/Table.mdx:99]: Please verify if the "Component API" table covers all props from the interface

File: packages/react-components/src/components/Table/Table.module.scss [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+  &--pinned_header {
+    @include pin.generatePin('header');
+  }
+
+  &__cell {
+    text-align: start;
+    line-height: 21px;
+    font-size: 14px;
+  }
```
[the remark this answers is not in the document]
Comment (marcinsawicki) [reply packages/react-components/src/components/Table/Table.module.scss:37]: To verify if it is possible to use our typography in the cell instead styling it

File: packages/react-components/src/components/Table/Table.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      <Table data={data} columns={columns} getRowId={getRowId} />
+    );
+
+    const table = getByRole('table');
+    expect(table).toBeInTheDocument();
+
+    const rows = getAllByRole('row');
+    expect(rows.length).toBe(data.length + 1);
+  });
+
```
[the remark this answers is not in the document]
Comment (marcinsawicki) [reply packages/react-components/src/components/Table/Table.spec.tsx:40]: We verify styling using Chromatic, and our eyes :) Not needed to cover it by the test

File: packages/react-components/src/components/Table/Table.tsx [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    };
+
+    const handleMouseUp = () => {
+      document.removeEventListener('mousemove', handleMouseMove);
+      document.removeEventListener('mouseup', handleMouseUp);
+    };
+
+    document.addEventListener('mousemove', handleMouseMove);
+    document.addEventListener('mouseup', handleMouseUp);
+  };
```
[the remark this answers is not in the document]
Comment (marcinsawicki) [reply packages/react-components/src/components/Table/Table.tsx:91]: Please verify this comment

**Comment by marcinsawicki** [comment]
Great job with this component. 💪 
This will be a good starting point for future improvements of `Table`.
Please verify the comments and implement changes based on the Vlad UI verification.

**Comment by vladko-uxds** [comment]
@arturfracala I see that when column is Stripped, the header still without it, the whole column (Cells + header) should have same background

https://github.com/user-attachments/assets/21b77757-906f-45bd-85ba-d0fd4cd6acad

**Comment by vladko-uxds** [comment]
@arturfracala when column in pinned  and we have a hover on a row - the whole row should have hover effect pinned cell too

Uploading hover with pinned column.mp4…

**Comment by vladko-uxds** [comment]
@arturfracala For resize columns please increase line width -> 2px
resize line

File: packages/react-components/src/components/Table/Table.mdx (hunk 3 above)
[the remark this answers is not in the document]
Comment (arturfracala) [reply packages/react-components/src/components/Table/Table.mdx:101]: All props are covered.

File: packages/react-components/src/components/Table/Table.module.scss [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    @include size.table-size(56px, 12px, 16px);
+  }
+
+  &--large {
+    @include size.table-size(64px, 14px, 16px);
+  }
+
+  &--pinned_header {
+    @include pin.generatePin('header');
+  }
```
[the remark this answers is not in the document]
Comment (arturfracala) [reply packages/react-components/src/components/Table/Table.module.scss:27]: It uses `--shadow-fixed-bottom`. It generates from function `pin.generatePin`

File: packages/react-components/src/components/Table/Table.mdx (hunk 2 above)
Comment (arturfracala) [reply packages/react-components/src/components/Table/Table.mdx:19]: Done.

**Review by vladko-uxds — APPROVED** [review]
Thanks for the huge contribution 👍

**Pull request merged**